ciongke/gcs/gcs.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcs
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	storage "google.golang.org/api/storage/v1"
+)
+
+// These will be reported in finished.json.
+const (
+	Success  = "SUCCESS"
+	Unstable = "UNSTABLE"
+	Failure  = "FAILURE"
+	Aborted  = "ABORTED"
+	Unknown  = "UNKNOWN"
+)
+
+const (
+	scope       = storage.DevstorageFullControlScope
+	contentType = "text/plain"
+
+	buildNumRetries = 3
+)
+
+// UploadSource uploads the path to the bucket.
+func UploadSource(bucket, name, path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return UploadToGCS(f, bucket, name+".tar.gz")
+}
+
+// IncrementAndGetBuildNumber increments and returns the build number stored in
+// gs://bucket/jobName/latest-build.txt. If the file does not exist, it sets it
+// to 0. Otherwise, it tries to be smart with object generation to avoid races.
+func IncrementAndGetBuildNumber(bucket, jobName string) (int, error) {
```

Comment:
What do you think about a monotonically increasing number (aka timestamp) over an incremental one?